### PR TITLE
Add circuit dag

### PR DESCRIPTION
**Pull request by mlodi-hqs**

### CODE REVIEW

File: roqoqo/tests/integration/circuitdag.rs [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    let ind1 = dag.add_to_back(operation1.clone());
+    let ind2 = dag.add_to_back(operation2.clone());
+
+    assert!(dag
+        .graph()
+        .contains_edge(ind1.unwrap().into(), ind2.unwrap().into()));
+}
+
```
Comment (nfwvogt) [inline roqoqo/tests/integration/circuitdag.rs:90]: Could you also add a test for edges if we have a all-qubit-operation in the circuit as this is a special case?

**Review by nfwvogt — CHANGES_REQUESTED** [review]
Looks good overall, but some changes could make the API more general.

Also for testing, once the reverse `From<CircuitDag> for Circuit` is implemented we should add a round-trip test with Circuit -> CircuitDag -> Circuit with a random circuit containing CNOT and PauliX. The result of the circuit can be calculated classically and we can compare the result of the original circuit and the round trip circuit.

File: roqoqo/tests/integration/circuitdag.rs [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+#[test_case(Operation::from(PauliX::new(0)), Operation::from(PauliY::new(0)))]
+fn check_edge(operation1: Operation, operation2: Operation) {
+    let mut dag: CircuitDag = CircuitDag::new();
+
+    let ind1 = dag.add_to_back(operation1.clone());
+    let ind2 = dag.add_to_back(operation2.clone());
+
+    assert!(dag
+        .graph()
+        .contains_edge(ind1.unwrap().into(), ind2.unwrap().into()));
```
[the remark this answers is not in the document]
Comment (mlodi-hqs) [reply roqoqo/tests/integration/circuitdag.rs:88]: They are Option<usize>, so they need it.

**Review by nfwvogt — APPROVED** [review]
Looks good to me

**Pull request merged**